### PR TITLE
[R] Fix nthread in DMatrix constructor.

### PR DESCRIPTION
**Pull request by trivialfis**
Close https://github.com/dmlc/xgboost/issues/7098 .

Restricting the construction of DMatrix to be single threaded.

### CODE REVIEW

**Comment by hcho3** [comment]
Is it possible to make `XGDMatrixCreateFromMat_R` aware of `nthread` setting?

**Comment by trivialfis** [comment]
That's either a new API or a breaking change.

**Comment by hcho3** [comment]
@trivialfis I don't think changing the signature for `XGDMatrixCreateFromMat_R` would break any user-facing functionality. We'd just need to modify `xgb.DMatrix.R`. What do you think?

**Comment by trivialfis** [comment]
I was told that the C function in R is also considered as part of the interface when working on https://github.com/dmlc/xgboost/pull/6819 .  That's why the old predict C function for R is still here.

**Review by RAMitchell — APPROVED** [review]

File: R-package/src/xgboost_R.cc [hunk 1]
```diff
@@ -103,7 +103,7 @@ XGB_DLL SEXP XGDMatrixCreateFromMat_R(SEXP mat, SEXP missing) {
   }
```
Comment (hcho3) [inline R-package/src/xgboost_R.cc:103]: The for loop in lines 96-102 is still using multiple threads. We must revise the loop also.

**Review by hcho3 — CHANGES_REQUESTED** [review]

**Comment by hcho3** [comment]
Also, let's create a new API function, `XGDMatrixCreateFromMat_omp_R`, that takes in the extra parameter `nthread`, so that R users do not lose out on the ability to use multiple threads when loading from a matrix.

File: R-package/src/xgboost_R.cc (hunk 1 above)
Comment (trivialfis) [reply R-package/src/xgboost_R.cc:103]: It's actually using the default.

> XGDMatrixCreateFromMat_omp_R, that takes in the extra parameter nthread

Maybe a new JSON based API.

Comment (hcho3) [reply R-package/src/xgboost_R.cc:103]: "The default" is set by `OMP_NUM_THREADS`, not `nthread.

**Comment by hcho3** [comment]
Let's just change the signature for `XGDMatrixCreateFromMat_R`, since this function is only used by the R package.

**Review by RAMitchell — APPROVED** [review]

**Pull request merged**